### PR TITLE
feature: add execution timeout for remote UDFs

### PR DESCRIPTION
**Pull request by amamiya-len**
PR checklist:
- Did you run ClangFormat ? Yes
- Did you separate headers to a different section in existing community code base ? Yes
- Did you surround `proton: starts/ends` for new code in existing community code base ? Yes

Please write user-readable short description of the changes:
enable execution timeout setting to close #5898

### CODE REVIEW

File: src/Parsers/ParserCreateFunctionQuery.cpp [hunk 1]
```diff
@@ -167,6 +169,20 @@ bool ParserCreateFunctionQuery::parseImpl(IParser::Pos & pos, ASTPtr & node, Exp
                 throw Exception("AUTH_METHOD must be 'none' or 'auth_header'", ErrorCodes::UNKNOWN_FUNCTION);
             }
         }
+        else
```
Comment (sunset3000) [inline src/Parsers/ParserCreateFunctionQuery.cpp]: remove the 'else'

File: src/Parsers/ASTCreateFunctionQuery.cpp [hunk 2]
```diff
@@ -81,6 +81,7 @@ void ASTCreateFunctionQuery::formatImpl(const IAST::FormatSettings & settings, I
             settings.ostr << fmt::format("AUTH_HEADER '{}'\n", function_core->children[1]->as<ASTLiteral>()->value.safeGet<String>());
             settings.ostr << fmt::format("AUTH_KEY '{}'\n", function_core->children[2]->as<ASTLiteral>()->value.safeGet<String>());
         }
+        settings.ostr << fmt::format("EXECUTION_TIMEOUT {}", function_core->children.back()->as<ASTLiteral>()->value.safeGet<UInt64>());
```
Comment (sunset3000) [inline src/Parsers/ASTCreateFunctionQuery.cpp]: What if the user does not set 'command_execution_timeout'?

File: src/Functions/UserDefined/UserDefinedFunctionConfiguration.h [hunk 3]
```diff
@@ -80,6 +80,8 @@ struct RemoteUserDefinedFunctionConfiguration : public UserDefinedFunctionConfig
     /// Timeout for reading data from input format
     size_t command_read_timeout_milliseconds = 10000;
 
+    size_t command_execution_timeout_milliseconds = 60000;
```
Comment (sunset3000) [inline src/Functions/UserDefined/UserDefinedFunctionConfiguration.h]: default time out should be 10000

File: src/Parsers/ASTCreateFunctionQuery.cpp (hunk 2 above)
Comment (amamiya-len) [reply src/Parsers/ASTCreateFunctionQuery.cpp]: the default value 10000 will be inserted and It is ensured that children of function_core is always at least 2( the first one is always auth_method and the last one is always execution timeout )

File: src/Parsers/ParserCreateFunctionQuery.cpp (hunk 1 above)
Comment (amamiya-len) [reply src/Parsers/ParserCreateFunctionQuery.cpp]: This can ensure that the first one of url's children is always AUTH_METHOD.

File: src/Common/sendRequest.cpp [hunk 4]
```diff
@@ -33,10 +33,10 @@ std::pair<String, Int32> sendRequest(
     const String & password,
     const String & payload,
     const std::vector<std::pair<String, String>> & headers,
-    Poco::Logger * log)
-{
+    Poco::Logger * log,
     /// One second for connect/send/receive
-    ConnectionTimeouts timeouts({2, 0}, {5, 0}, {10, 0});
+    ConnectionTimeouts timeouts)
```
Comment (sunset3000) [inline src/Common/sendRequest.cpp]: the last param is 'log', move the 'timeouts' param up.

File: src/Common/sendRequest.cpp [hunk 5]
```diff
@@ -109,6 +109,18 @@ std::pair<String, Int32> sendRequest(
         {
             session->attachSessionData(e.message());
         }
+        if (e.code() == 1000){
+            LOG_ERROR(
+                log,
+                "Execution Timeout from uri={} method={} payload={} query_id={} error={} exception={}",
```
Comment (sunset3000) [inline src/Common/sendRequest.cpp]: change 'Timeout' to 'timeout'

File: src/Functions/UserDefined/RemoteUserDefinedFunction.h [hunk 6]
```diff
@@ -53,7 +54,8 @@ class RemoteUserDefinedFunction final : public UserDefinedFunctionBase
             "",
             out,
             {{config.auth_context.key_name, config.auth_context.key_value}, {"", context->getCurrentQueryId()}},
-            &Poco::Logger::get("UserDefinedFunction"));
+            &Poco::Logger::get("UserDefinedFunction"),
+            ConnectionTimeouts({2, 0}, {5, 0}, {static_cast<long>(config.command_execution_timeout_milliseconds / 1000), static_cast<long>((config.command_execution_timeout_milliseconds % 1000u) * 1000u)}));
```
Comment (sunset3000) [inline src/Functions/UserDefined/RemoteUserDefinedFunction.h]: add comments for 'ConnectionTimeouts'
```C++
ConnectionTimeouts({2, 0}, {5, 0}, {static_cast<long>(config.command_execution_timeout_milliseconds / 1000), static_cast<long>((config.command_execution_timeout_milliseconds % 1000u) * 1000u)})) ///  timeout and limit for connect/send/receive/tcp_keep_alive/http_keep_alive ...
```

File: src/Parsers/ASTCreateFunctionQuery.cpp (hunk 2 above)
Comment (sunset3000) [reply src/Parsers/ASTCreateFunctionQuery.cpp]: Who and how is the default value inserted if not provided in SQL?

File: src/Parsers/ParserCreateFunctionQuery.cpp (hunk 1 above)
Comment (sunset3000) [reply src/Parsers/ParserCreateFunctionQuery.cpp]: if do not remvoe 'else' the result is incorrect.

File: src/Parsers/tests/gtest_create_remote_func_parser.cpp
Comment (sunset3000) [inline src/Parsers/tests/gtest_create_remote_func_parser.cpp:1]: add more tests for the corner cases, like auth_method is not provided and only execution_timeout has been provided, or no one has been provided, et al.

File: src/Common/sendRequest.cpp (hunk 4 above)
Comment (amamiya-len) [reply src/Common/sendRequest.cpp]: `timeouts` is orginally a local variable in this function. This change will maximize compatibility for other codes using `sendRequest`

Comment (amamiya-len) [reply src/Common/sendRequest.cpp]: I have moved the `timeout` param up.

File: src/Functions/UserDefined/UserDefinedFunctionConfiguration.h (hunk 3 above)
Comment (amamiya-len) [reply src/Functions/UserDefined/UserDefinedFunctionConfiguration.h]: I have modify it to 10000

**Comment by amamiya-len** [comment]
TODO: 
- [x] change the way to store the props in  `ASTRemoteFunction`.
- [x] add more unit tests

File: src/Coordination/MetaStoreConnection.cpp [hunk 7]
```diff
@@ -215,7 +215,7 @@ std::pair<String, Int32> MetaStoreConnection::forwardRequest(
     for (auto it = active_servers.begin(); it != active_servers.end();)
     {
         Poco::URI uri{fmt::format({METASTORE_URL}, it->getHost(), it->getPort(), uri_parameter)};
-        auto [response, http_status] = sendRequest(uri, method, query_id, user, password, body, {}, log);
+        auto [response, http_status] = sendRequest(uri, method, query_id, user, password, body, {}, ConnectionTimeouts({2, 0}, {5, 0}, {10, 0}) , log);
```
Comment (sunset3000) [inline src/Coordination/MetaStoreConnection.cpp]: add comment for 'ConnectionTimeouts'

File: src/Common/sendRequest.cpp [hunk 8]
```diff
@@ -33,10 +33,10 @@ std::pair<String, Int32> sendRequest(
     const String & password,
     const String & payload,
     const std::vector<std::pair<String, String>> & headers,
+    /// One second for connect/send/receive
```
Comment (sunset3000) [inline src/Common/sendRequest.cpp]: change to 'Timeout for connect/send/receive'

File: src/Functions/UserDefined/RemoteUserDefinedFunction.h [hunk 9]
```diff
@@ -53,6 +54,7 @@ class RemoteUserDefinedFunction final : public UserDefinedFunctionBase
             "",
             out,
             {{config.auth_context.key_name, config.auth_context.key_value}, {"", context->getCurrentQueryId()}},
+            ConnectionTimeouts({2, 0}, {5, 0}, {static_cast<long>(config.command_execution_timeout_milliseconds / 1000), static_cast<long>((config.command_execution_timeout_milliseconds % 1000u) * 1000u)}),
```
Comment (sunset3000) [inline src/Functions/UserDefined/RemoteUserDefinedFunction.h]: add comments for 'ConnectionTimeouts'

File: src/Functions/UserDefined/UserDefinedFunctionConfiguration.h [hunk 10]
```diff
@@ -80,6 +80,8 @@ struct RemoteUserDefinedFunctionConfiguration : public UserDefinedFunctionConfig
     /// Timeout for reading data from input format
     size_t command_read_timeout_milliseconds = 10000;
 
+    size_t command_execution_timeout_milliseconds = 10000;
```
Comment (sunset3000) [inline src/Functions/UserDefined/UserDefinedFunctionConfiguration.h:84]: add comments for this parameter

File: src/Parsers/tests/gtest_create_remote_func_parser.cpp [hunk 11]
```diff
@@ -112,5 +117,19 @@ TEST(ParserCreateRemoteFunctionQuery, UDFHeaderMethodIsOther)
...
+                   "AUTH_HEADER 'auth'"
+                   "AUTH_KEY 'proton'"
+                   "EXECUTION_TIMEOUT 30000"
+                   " AUTH_METHOD 'auth_header'";
+    ParserCreateFunctionQuery parser;
     EXPECT_THROW(parseQuery(parser, input.data(), input.data() + input.size(), "", 0, 0), Exception);
-}
\ No newline at end of file
+}
+
```
Comment (sunset3000) [inline src/Parsers/tests/gtest_create_remote_func_parser.cpp:183]: Need to add test cases to 
1) check ‘formatImpl’ of CreateFunctionQuery (remote udf)
2) force-error tests for missing required param, unsupportted param, unknown param et al.

File: src/Parsers/ASTCreateFunctionQuery.cpp [hunk 12]
```diff
@@ -153,20 +148,41 @@ Poco::JSON::Object::Ptr ASTCreateFunctionQuery::toJSON() const
...
+                    auth_header = pair->second->as<ASTLiteral>()->value.safeGet<String>();
+                else if (pair->first == "auth_key")
+                    auth_key = pair->second->as<ASTLiteral>()->value.safeGet<String>();
+                else if (pair->first == "execution_timeout")
+                    execution_timeout = pair->second->as<ASTLiteral>()->value.safeGet<UInt64>();
             }
+        }        
+
+        inner_func->set("url", url.value());
+        if (auth_method.has_value() && auth_method.value() == "auth_header")
```
Comment (sunset3000) [inline src/Parsers/ASTCreateFunctionQuery.cpp]: When auth_method == 'None', the 'auth_method' should be set to 'none'. instead of ignore the all 'auth_context' property.

File: src/Parsers/ParserCreateFunctionQuery.cpp [hunk 13]
```diff
@@ -139,35 +143,10 @@ bool ParserCreateFunctionQuery::parseImpl(IParser::Pos & pos, ASTPtr & node, Exp
         {
             throw Exception("Remote udf can not be an aggregate function", ErrorCodes::AGGREGATE_FUNCTION_NOT_APPLICABLE);
         }
-        if (!s_url.ignore(pos, expected))
+        ParserKeyValuePairsSet kv_pairs_list;
+        if (!kv_pairs_list.parse(pos, kv_list, expected))
```
Comment (sunset3000) [inline src/Parsers/ParserCreateFunctionQuery.cpp:147]: Need to check the required parameters for Remote UDF.

File: src/Parsers/ParserKeyValuePairsSet.h [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    ParserPtr separator_parser = std::make_unique<ParserNothing>();
+    bool allow_empty = true;
+    char result_separator = '\0';
+    const char * getName() const override { return "set of pairs"; }
+    bool parseImpl(Pos & pos, ASTPtr & node, Expected & expected, [[ maybe_unused ]] bool hint) override;
+};
+
+}
+
+
```
Comment (sunset3000) [inline src/Parsers/ParserKeyValuePairsSet.h]: remove the extra blank line

**Review by sunset3000 — APPROVED** [review]
